File: Neetcode Questions/Math & Geometry/spiral-matrix.cpp

// https://leetcode.com/problems/spiral-matrix/

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int r = matrix.size();
        if(r == 0) return ans;
        int c = matrix[0].size();

        int total = r*c;
        int count = 0;
        int startingRow = 0, endingRow = r-1;
        int startingCol = 0, endingCol = c-1;

        while(count < total){
            for(int i = startingCol; i <= endingCol && count < total; i++, count++) ans.push_back(matrix[startingRow][i]);
            startingRow++;
            for(int i = startingRow; i <= endingRow && count < total; i++, count++) ans.push_back(matrix[i][endingCol]);
            endingCol--;
            for(int i = endingCol; i >= startingCol && count < total; i--, count++) ans.push_back(matrix[endingRow][i]);
            endingRow--;
            for(int i = endingRow; i >= startingRow && count < total; i--, count++) ans.push_back(matrix[i][startingCol]);
            startingCol++;
        }

        return ans;
    }
};

File: Neetcode-150-js/Math & Geometry/spiral-matrix.js

// https://leetcode.com/problems/spiral-matrix/description/

var spiralOrder = function(matrix) {

    const ans = [];

    let rowSize = matrix.length;
    if(rowSize === 0) return ans;
    let colSize = matrix[0].length;
    
    let totalElements = rowSize*colSize-1;
    let startingRow = 0;
    let endingRow = rowSize-1;
    let startingCol = 0;
    let endingCol = colSize-1;
    let k = 0;

    while(k <= totalElements){
        for(let i = startingCol; i <= endingCol && k <= totalElements; i++){
            ans.push(matrix[startingRow][i]);
            k++;
        }
        startingRow++;

        for(let i = startingRow; i <= endingRow && k <= totalElements; i++){
            ans.push(matrix[i][endingCol]);
            k++;
        }
        endingCol--;

        for(let i = endingCol; i >= startingCol && k <= totalElements; i--){
            ans.push(matrix[endingRow][i]);
            k++;
        }
        endingRow--;

        for(let i = endingRow; i >= startingRow && k <= totalElements; i--){
            ans.push(matrix[i][startingCol]);
            k++;
        }
        startingCol++;
    }

    return ans;
};